src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1018,17 +990,44 @@ DatabaseShardImp::sweep()
         assert(init_);
 
         for (auto const& e : shards_)
-            if (e.second.state == ShardInfo::State::final ||
-                e.second.state == ShardInfo::State::acquire)
-            {
-                shards.push_back(e.second.shard);
-            }
+            shards.push_back(e.second);
     }
 
+    std::vector<std::shared_ptr<Shard>> openFinals;
     for (auto const& e : shards)
     {
         if (auto shard{e.lock()}; shard)
-            shard->sweep();
+        {
+            if (shard->isOpen())
+            {
+                shard->sweep();
+
+                if (shard->getState() == Shard::State::final)
+                    openFinals.emplace_back(std::move(shard));
+            }
+        }
+    }
+
+    if (openFinals.size() > openFinalLimit_)
+    {
+        // Try to close enough shards to be within the limit.
```

Comment:
Do you think it's worth it to provide trace or debug logging on the number of open shards vs the configured limit?

Reply:
Good idea, will add.